### PR TITLE
Apply `Array#push`'s resizing heuristic to `#unshift`

### LINKED ISSUE #10748: Memory leak on pop followed by unshift
this was found testing against crystal v1.0.0
changing from inline sort to creating a new object resolved the issue so I assume there is an issue with `Slice.sort!`

`array.sort! { |a, b| a.test <=> b.test }` to `array = array.sort { |a, b| a.test <=> b.test }`

the code

```crystal
class Task
  def initialize
    @test = rand(10000)
  end
  getter test : Int32
end

array = [] of Task
loop do
  task = Task.new
  array.unshift task
  array.sort! { |a, b| a.test <=> b.test }
  puts "array size == #{array.size}"
  array.pop
end

```

the output, 28 calls to `sort!` results in about 28GB in memory usage

```
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
array size == 1
GC Warning: Repeated allocation of very large block (appr. size 805310464):
	May lead to memory leak and poor performance
array size == 1
array size == 1
array size == 1
Program was killed
```

### PR DESCRIPTION
**Pull request by HertzDevil**
Fixes #10748.

The following illustrates some examples of the heuristic in action:

```crystal
class Array(T)
  def to_s(io : IO) : Nil
    root = root_buffer
    io << '['
    @capacity.times do |i|
      io << ", " if i > 0
      if @offset_to_buffer <= i < @offset_to_buffer + @size
        root[i].inspect(io)
      else
        io << '_'
      end
    end
    io << ']'
  end
end
```

```crystal
x = Array(Int32).new(5)
x.push(1, 2); puts x
x.unshift(3); puts x
x.push(4); puts x

# before:
# [1, 2, _, _, _]
# [_, _, _, _, 3, 1, 2, _, _, _]
# [_, _, _, _, 3, 1, 2, 4, _, _]

# after:
# [1, 2, _, _, _]
# [_, 3, 1, 2, _] # there is enough space; move 1 and 2 to the middle of the root buffer
# [_, 3, 1, 2, 4]
```

```crystal
x = Array(Int32).new(4)
x.unshift(0); puts x
x.shift; x.unshift(1); puts x
x.shift; x.unshift(2); puts x
x.unshift(4, 3); puts x

# before:
# [_, _, _, 0, _, _, _, _]
# [_, _, _, _, _, _, _, 1, _, _, _, _, _, _, _, _]
# [_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, 2, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _]
# [_, _, _, _, _, _, _, _, _, _, _, _, _, 4, 3, 2, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _]

# after:
# [_, 0, _, _] # there is pre-allocated space; start inserting elements from the middle
# [_, 1, _, _]
# [_, 2, _, _]
# [_, 4, 3, 2] # there is enough space; move 3 and 2 to the middle of the root buffer
```

### CODE REVIEW

**Review by stakach — APPROVED** [review]
LGTM

**Comment by stakach** [comment]
Hopefully this can make it into `v1.0.1`

File: spec/std/array_spec.cr [hunk 1]
```diff
@@ -1497,6 +1497,33 @@ describe "Array" do
       a.@offset_to_buffer.should eq(0)
       a.@buffer.should eq(buffer)
     end
+
+    it "repeated unshift/shift does not exhaust memory" do
+      a = [] of Int32
+      10000.times do
```
Comment (straight-shoota) [inline spec/std/array_spec.cr:1503]: Do such many iterations bring any reasonable benefit?
I figure even two passes would be sufficient proof. Or make it three to be safe.

**Review by straight-shoota — APPROVED** [review]
This is probably fine for a quick fix.

But I think it might be beneficial to take a closer look at the applied heuristics. I could see some room for improvement there.
A worst-case scenario with this is adding an element each to the front and back of an array that's filled to capacity (a typical example would be literals). Adding just these two elements quadruples the capacity: it's doubled by both push and unshift, regardless of their order. This means allocating a lot of unnecessary memory.
If the resize operations for push and unshift would not fill the buffer all the way to the front and back, respectively, this could be avoided.
I don't know what would be a better solution and I'm not even sure this needs to be changed, but it probably deserves an evaluation and discussion. Obviously, the best solution depends on a fair balance between different use cases.

**Comment by asterite** [comment]
We should probably merge this since it's an important fix

**Review by beta-ziliani — APPROVED** [review]
Let's get this for 1.1.0. There's the unanswered question about the need for 10k iterations in tests, when probably a handful ones should suffice. But we can change that later (I'd like to hear @HertzDevil opinion on the matter).

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**